### PR TITLE
[Console] Add LockableInterface and automatic locking when it's implemented

### PR DESCRIPTION
**Pull request by j4nr6n**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        | WIP

This PR adds automatic locking for commands that implement `Symfony\Component\Console\Command\LockableInterface`. It is currently a work in progress. I'm creating the PR now to, hopefully, get some feedback before I continue much further.

* [ ] Tests
* [ ] Docs

### CODE REVIEW

**Comment by ostrolucky** [comment]
In final version, I hope to see dependency injection to be used

**Comment by j4nr6n** [comment]
It was heavily based on `LockableTrait::lock` which doesn't use dependency injection. I was able to clean up the code a bit by using `LockableTrait`.

File: src/Symfony/Component/Console/Application.php [hunk 1]
```diff
@@ -923,6 +928,14 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
         $e = null;
 
         try {
+            if (class_exists(Lock::class) && \in_array(LockableTrait::class, class_uses($command))) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Application.php:931]: I think all `class_exists(Lock::class)` checks should be removed.

File: src/Symfony/Component/Console/Application.php [hunk 2]
```diff
@@ -923,6 +928,14 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
...
+                $command->getDefinition()->addOption(new InputOption('--skip-lock', '', InputOption::VALUE_NONE, 'Skip checking for and setting a lock'));
+
+                if (false === $input->hasParameterOption(['--skip-lock'], true) && false === $this->lock($command->getName().'_symfony')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Application.php:934]: the lock should be released() somewhere

**Review by nicolas-grekas — COMMENTED** [review]
I'm not sure using the trait as a trigger is the correct design. E.g. what if one needs a distributed lock (aka not the implementation provided by LockableTrait)?
Adding a LockableInterface interface might be more flexible.

**Comment by j4nr6n** [comment]
@nicolas-grekas The funny thing is, I made that change but I wasn’t sure so I backed it out. 😆

File: src/Symfony/Component/Console/Application.php (hunk 2 above)
Comment (j4nr6n) [reply src/Symfony/Component/Console/Application.php]: Agreed. I added a rough release mechanism for now. Though, I'm wondering if both of these parts can be moved into event listeners now that the methods are defined in an interface.

**Comment by j4nr6n** [comment]
Still a bit confused on the implementation of the `LockableInterface`, you can't `use LockableTrait` to implement the interface because the methods in `LockableTrait` are private. Should that be changed?

**Comment by j4nr6n** [comment]
> we've implemented this via events

I'm going to attempt that change.

> also no fan of the global skip-lock option, a contract seems more appropriate

It's not a global option. In it's current form, it's only added to commands that implement `LockableInterface`.

**Comment by ro0NL** [comment]
global behavior for all lockables then :}

we needed e.g. 

```
    public function shouldLock(InputInterface $input): bool
    {
        return !$input->getOption('dry-run') && null !== $input->getOption('lock-ttl');
    }

    public function getLockTtl(InputInterface $input): ?float
    {
        return (float) $input->getOption('lock-ttl');
    }
```

to solve  a real life usecase.

**Comment by j4nr6n** [comment]
I'd prefer to keep things in line with `LockableTrait` so that the required methods can be implemented with `use LockableTrait;`. Since the methods are defined in an interface, there is no way to define a body or default return value that I'm aware of. I don't want to force implementation of additional methods and I'm not sure `shouldLock` and `getLockTtl` should be included in `LockableTrait`.

**Comment by ro0NL** [comment]
at least, IMHO the Command is in charge of resource locking; not the Application in some generic way. Thus the command may want to decide whether it should lock yes/no, its resource name and/or its ttl. Even on a per signature basis, as i've showed an example of.

**Comment by ro0NL** [comment]
to evaluate my own implementation even further, i'd probably go with

```
getLockResource(InputInterface $input): ?LockResource {
  return new LockResource(?string $name, ?int $tll, ?StoreInterface $store)
}
```

**Comment by j4nr6n** [comment]
The command is still in charge of locking, completely. This PR adds logic to call the `lock` method if the new `LockableInterface` is implemented (maybe a more clear name can be used, maybe `AutoLockableInterface`?). It doesn't change anything else. The command name is appended with `_symfony` (E.G. `app:test_symfony`) for the lock name, specifically to differentiate it from locks that people may have already implemented using the command name. `LockableTrait::lock` doesn't currently implement TTL or the logic to decide if the command should lock. If `LockableTrait::lock` isn't what you need, you can still override it or implement your own `lock` method where you can perform any sort of logic you want.

The purpose of this PR is to make basic locking easier, rather than implementing the locking logic in every command class or overriding `initialize` in some custom `AbstractCommand` class to add the logic to every custom command and then having to modify `bin/console` to add a global option to all commands (including built-in commands that don't use it).

I'm not understanding your examples. I don't see how or where they fit into these changes. To me, all of it seems like logic you would put inside of `lock()` or protected/private methods in your command that you would call from `lock()`.

**Comment by ro0NL** [comment]
oh right. i forgot Command::lock still encapsulates the actual locking. Fair, i dont see any practical downsides :+1:

File: src/Symfony/Component/Console/Application.php [hunk 3]
```diff
@@ -923,6 +925,14 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
         $e = null;
 
         try {
+            if ($command instanceof LockableInterface) {
+                $command->getDefinition()->addOption(new InputOption('--skip-lock', '', InputOption::VALUE_NONE, 'Skip checking for and setting a lock'));
+
+                if (false === $input->hasParameterOption(['--skip-lock'], true) && false === $command->lock($command->getName().'_symfony')) {
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Application.php:931]: i dont think it should pass a $name actually, the null case is a nice extension point for commands to define a name with more variance

File: src/Symfony/Component/Console/Application.php [hunk 4]
```diff
@@ -947,6 +957,10 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
             throw $e;
         }
 
+        if ($command instanceof LockableInterface) {
+            $command->release();
```
Comment (ro0NL) [inline src/Symfony/Component/Console/Application.php:961]: shouldnt it be released before the throw...

**Review by ro0NL — COMMENTED** [review]
not convinced on `--skip-lock`, i tend to believe it should be a Command::lock implementation detail

still a bit bummed it cant easily switch to e.g. a RedisStore :(

File: src/Symfony/Component/Console/Application.php (hunk 3 above)
Comment (j4nr6n) [reply src/Symfony/Component/Console/Application.php:931]: It's a default value, you can totally override it. See my example implementation of this feature below.

**Comment by j4nr6n** [comment]
You _can_ use a RedisStore, or any store for that matter as well as any logic you'd like to use.
<details>
    <summary>Example</summary>

    class AutomaticLockingCommand extends Command implements LockableInterface
    {
        protected static $defaultName = 'app:automatic-locking';

        /** @var Lock */
        private $lock;

        /** @var LockFactory */
        private $lockFactory;

        public function __construct(LockFactory $lockFactory)
        {
            $this->lockFactory = $lockFactory;

            // You must call the parent constructor.
            parent::__construct();
        }

        protected function configure()
        {
            // ...
        }

        protected function execute(InputInterface $input, OutputInterface $output)
        {
            // ...
        }

        public function lock(string $name = null, bool $blocking = false): bool
        {
            if (!$this->shouldLock()) {
                // Don't lock
                return true;
            }

            $lockTTL = 30;
            $this->lock = $this->lockFactory->createLock('some_custom_name', $lockTTL);

            return $this->lock->acquire();
        }

        public function release()
        {
            if ($this->lock) {
                $this->lock->release();
                $this->lock = null;
            }
        }

        private function shouldLock()
        {
            // ...
        }
    }
</details>

And if you don't want command level, automatic locking, don't add the `implements LockableInterface` to your commands. Things will stay the same as they are now.

File: src/Symfony/Component/Console/Application.php (hunk 4 above)
Comment (j4nr6n) [reply src/Symfony/Component/Console/Application.php:957]: Your right! Fixed in latest push.

File: src/Symfony/Component/Console/Application.php (hunk 3 above)
Comment (ro0NL) [reply src/Symfony/Component/Console/Application.php:931]: i understand, but ignoring a user provided $name cant be the right implementation IMHO.

Id like to do something like `$name = $name ?? 'my-default-command-lock-name'`

**Comment by ro0NL** [comment]
@j4nr6n you're right. For this purpose `LockableTrait::getLockStore()` sounds like a nice extension point. WDYT.

**Comment by j4nr6n** [comment]
What is the purpose? This feature does not care what lock store you use or what name you want to give the lock. That is all implementation detail and that is entirely up to you. All this code does is call `$command->lock()` and `$command->release()` if the command class is an instance of `LockableInterface`. The examples I gave are directly based on the Lock component’s existing features and documentation. I’m not changing anything about how the lock component currently works.

**Pull request closed without merging**